File: cpp/tennisKata/src/Tennis.cpp
/*
 * =====================================================================================
 *       Filename:  Tennis.cpp
 *        Created:  2014-06-26 22:52
 *         Author:  Damian Ziobro        (ziobro.damian@gmail.com)
 *        Company:  XMementoIT Limited       (info@xmementoit.com)
 * =====================================================================================
 */

#include "Tennis.h"

const string Tennis::POINT_NAMES[4] = {"love", "fifteen", "thirty", "fourty"};

/* ======= Function ==================================================
 *   Name: Tennis::Tennis
 *   Description: Default contructor of class Tennis 
 * =================================================================== 
 */
Tennis::Tennis(shared_ptr<Player>& player1, shared_ptr<Player>& player2)
{
    this->player1 = player1;
    this->player2 = player2;
}

/* ======= Function ==================================================
 *   Name: Tennis::~Tennis
 *   Description: Destructor of class Tennis
 * =================================================================== 
 */
Tennis::~Tennis()
{
    //TODO
}

unique_ptr<string> Tennis::sayResult() {

    unique_ptr<string> result(new string(""));

    int pts1 = player1->getPoints();
    int pts2 = player2->getPoints();

    if (pts1 >= 3 && pts2 >= 3){
      //advantage case
      if (pts1 == pts2 +1)  {
        *result = "advantage " + player1->getName();
      } else if (pts2 == pts1 + 1) {
        *result = "advantage " + player2->getName();
      } else if(pts1 == pts2) {
        *result = "deuce";
      } else {
        if (pts1 > pts2) {
          *result = "Won " + player1->getName();
        } else {
          *result = "Won " + player2->getName();
        }
      }
    } else {
      *result = POINT_NAMES[player1->getPoints()] + " " + \
               POINT_NAMES[player2->getPoints()];
    }

    return result;
}

File: java/tennis/src/main/java/com/company/Tennis.java
/*
 * Tennis.java
 * Copyright (C) 2018 damian <damian@damian-laptop>
 *
 * Distributed under terms of the MIT license.
 */

package com.company;

import com.company.ITennis;
import java.lang.IllegalArgumentException;
import java.lang.StringBuilder;

public class Tennis implements ITennis
{
  protected IPlayer player1;
  protected IPlayer player2;

  protected String[] results = new String[]{"love", "fifteen", "thirty", "forty"};

	public Tennis(IPlayer p1, IPlayer p2) {
    this.setPlayers(p1, p2);
	}

  public void setPlayers(IPlayer p1, IPlayer p2) throws IllegalArgumentException {
    if (p1 == p2) {
      throw new IllegalArgumentException();
    }

    this.player1 = p1;
    this.player2 = p2;
  }

  public String getResult(){
    StringBuilder result = new StringBuilder("");
    int pts1 = player1.getPoints();
    int pts2 = player2.getPoints();
    
    //test whether anyone won
    if (pts1 > 3 && pts1 >= pts2+2) {
      result.append("Won " + player1.getName());
    } else if (pts2 > 3 && pts2 >= pts1+2) {
      result.append("Won " + player2.getName());
    } else if (pts1 >= 3 && pts2 >= 3) {
      //check advantage or deuce
      if (pts1 == pts2) {
        result.append("deuce");
      } else {
        result.append("advantage ");
        if (pts1 == pts2+1) {
          result.append(player1.getName());
        } else if (pts2 == pts1 +1) {
          result.append(player2.getName());
        } 
      }
    } else {
      //normal result 
      result.append(results[pts1] + " " + results[pts2]);
    }


    return result.toString();
  }

}


